### ISSUE TITLE
Inject events via  [[UIApplication sharedApplication] _enqueueHIDEvent:]

### ISSUE THREAD

**Issue by steipete** [google/EarlGrey#293]
Digging into EarlGrey again and all the technical crazy/beauty that's in there. This is such an awesome project.

So. Did some research around the use of `IOHIDEvent` since both KIF and EarlGrey need to use that, and found out that WebKit's TestRunner is doing that as well - but slightly differently than the approach on our testing tools here.

Check out this line specifically:
https://github.com/WebKit/webkit/blob/master/Tools/WebKitTestRunner/ios/HIDEventGenerator.mm#L265

The whole code and handling is interesting and while I don't know what the difference yet, is, see this issue as a TODO to investigate more. Can't promise I'll find the time but maybe one of you wants to take a look if this might be useful.

**Comment by tirodkar** [comment]
Thanks for bringing this to our attention Peter! For future EarlGrey versions, we'd want to have the tapping handled directly by XCUITests's tap. If I'm right, that would pretty much circumvent the need to use `IOHIDEvent` altogether.

**Comment by khandpur** [comment]
Yup. One of the goals for the next version of EarlGrey is to reduce our dependency on private APIs with touch injection being the major contender.

**Comment by khandpur** [comment]
Actually I am going to give this a shot this week :)

**Comment by steipete** [comment]
Will watch your progress! 👌

**Comment by khandpur** [comment]
Update:
https://github.com/WebKit/webkit/blob/master/Tools/WebKitTestRunner/ios/HIDEventGenerator.mm#L265 doesn't work for all the supported platform versions. Also, it doesn't work on devices due to entitlement issues. And to top it off - there doesn't seem to be a reliable way to wait for the touch to be processes by the Application after it's been enqueued in the system. All of this makes it infeasible for EarlGrey. I have a "somewhat" cleaner solution that I am cleaning up right now and should post it on github by this week if I am able to get all the internal tests passing with it.
